src/charm.py

```diff
@@ -7,29 +7,176 @@
+
+        if not self._container.can_connect():
+            statuses.append(ops.WaitingStatus("Waiting for the Kiali container to be ready"))
+
+        if not (prometheus_source_available := self._is_prometheus_source_available()):
+            statuses.append(ops.BlockedStatus("Prometheus source is not available"))
+
+        if prometheus_source_available:
+            # Only valid if we have a prometheus source
+            kiali_config = self._generate_kiali_config()
+            kiali_local_url = f"http://localhost:{kiali_config['server']['port']}/kiali"
+            if not _is_kiali_available(kiali_local_url):
+                statuses.append(
+                    ops.WaitingStatus(
+                        "Kiali is configured and container is ready, but Kiali is not available"
+                    )
+                )
+
+        if len(statuses) == 0:
+            statuses.append(ops.ActiveStatus())
+
+        for status in statuses:
+            e.add_status(status)
+
+    def reconcile(self, _event: ops.ConfigChangedEvent):
         """Reconcile the entire state of the charm."""
-        self.unit.status = ops.BlockedStatus("todo: implement reconcile")
+        self._configure_kiali_workload()
+
+    def _configure_kiali_workload(self):
+        """Configure the Kiali workload, if possible, logging errors otherwise.
+
+        This will generate and push the Kiali configuration to the container, restarting the service if necessary.
+        The purpose here is that this should always attempt to configure/start Kiali, but it does not guarantee Kiali is
+        running after completion.  If any known errors occur, they will be logged and this method will return without
+        error.  To confirm if Kiali is working, check the status of the Kiali workload directly.
+        """
+        # TODO: Can we make this generic and share with other charms?
+        name = "kiali"
+        if not self._container.can_connect():
+            LOGGER.warning(f"Container is not ready, cannot configure {name}")
+            return
+
+        layer = self._generate_kiali_layer()
+        try:
+            new_config = yaml.dump(self._generate_kiali_config())
+        except GrafanaSourceError as e:
+            LOGGER.warning(f"Failed to generate {name} configuration, got error: {e}")
+            # TODO: actually shut down the service and remove the configuration
+            # LOGGER.warning(f"Shutting down {name} service and removing existing configuration")
+            return
+
+        should_restart = not _is_container_file_equal_to(
+            self._container, str(KIALI_CONFIG_PATH), new_config
+        )
+        self._container.push(KIALI_CONFIG_PATH, new_config, make_dirs=True)
+        self._container.add_layer(name, layer, combine=True)
+        self._container.autostart()
+
+        if should_restart:
+            LOGGER.info(f"new config detected for {name}, restarting the service")
+            self._container.replan()
```

Comment:
I think this should be `restart` rather than `replan`. I could be wrong but I think this only restarts the service if the *layer* changed. So if the layer stays the same and the config changes, we will not get a restart.

Reply:
Also, we should look into whether there is a "reload" functionality and if so, use that instead.

Reply:
I think your understanding of [replan](https://canonical-pebble.readthedocs-hosted.com/en/latest/reference/cli-commands/#replan) is correct - it only restarts the layer changed - and that you're correct that this should be [restart](https://canonical-pebble.readthedocs-hosted.com/en/latest/reference/cli-commands/#restart).  Good catch!

Changed